File: scripts/patch-ferry-runtime.mjs
import { readFileSync, writeFileSync } from 'node:fs';

const path = 'scripts/ferry-cloudflare-action.sh';
let source = readFileSync(path, 'utf8');

const sshSection = /mkdir -p "\$HOME\/\.ssh"[\s\S]*?ssh-keyscan -p 3022 localhost[^\n]*\n/;
if (!sshSection.test(source)) throw new Error('Could not locate Ferry SSH setup');
source = source.replace(
  sshSection,
  'docker exec dokku dokku network:set --global attach-post-deploy webserver || log "Dokku global network setting returned non-zero; continuing"\n',
);

const marker = 'log "Deploying the checked-out app through Ferry"';
if (!source.includes(marker)) throw new Error('Could not locate Ferry deploy marker');
source = source.replace(
  marker,
  `log "Initializing empty Cloudflare ingress before Ferry preflight"
ingress_payload='{"config":{"ingress":[{"service":"http_status:404"}]}}'
cf_result "$(cf PUT "/accounts/$CF_ACCOUNT_ID/cfd_tunnel/$TUNNEL_ID/configurations" "$ingress_payload")" >/dev/null

log "Creating deployment infrastructure with Ferry"`,
);

const deploySection = /export CF_EMAIL CF_GLOBAL_APIKEY[\s\S]*?\.\/ferry\.sh deploy "\$APP_NAME"[^\n]*\n\)/;
if (!deploySection.test(source)) throw new Error('Could not locate Ferry git-push deployment');
source = source.replace(
  deploySection,
  `export CF_EMAIL CF_GLOBAL_APIKEY CF_ACCOUNT_ID TUNNEL_ID TUNNEL_TOKEN DOKKU_HOSTNAME
export CF_API_TOKEN=global-key-compat
(
  cd "$RUNTIME"
  ./ferry.sh deploy "$APP_NAME" -H "$APP_HOSTNAME" -p "$APP_PORT" -d "$ROOT" --no-push -y
)

log "Building app image on the shared Docker daemon"
IMAGE_TAG="ferry-ci/$APP_NAME:${GITHUB_SHA:-latest}"
docker build --tag "$IMAGE_TAG" "$ROOT"
docker exec dokku dokku network:set "$APP_NAME" attach-post-deploy webserver || true
log "Releasing image through Dokku without SSH"
docker exec dokku dokku git:from-image "$APP_NAME" "$IMAGE_TAG"`,
);

writeFileSync(path, source);
console.log('Node.js patch applied: Ferry infrastructure + Dokku image release, no SSH');

File: scripts/patch-ferry-runtime.py
from pathlib import Path

path = Path("scripts/ferry-cloudflare-action.sh")
source = path.read_text()

source = source.replace(
    "docker exec dokku dokku network:set --global attach-post-deploy webserver\n",
    'docker exec dokku dokku network:set --global attach-post-deploy webserver || log "Dokku network setting returned non-zero; continuing"\n',
)

old_ssh = 'ssh-keyscan -p 3022 localhost >> "$HOME/.ssh/known_hosts" 2>/dev/null\n'
new_ssh = '''ssh_ready=false
for _ in $(seq 1 60); do
  if ssh-keyscan -p 3022 localhost >> "$HOME/.ssh/known_hosts" 2>/dev/null; then
    ssh_ready=true
    break
  fi
  sleep 2
done
$ssh_ready || die "Dokku SSH did not become ready"
'''
if old_ssh not in source:
    raise SystemExit("SSH patch target missing")
source = source.replace(old_ssh, new_ssh)

marker = 'log "Deploying the checked-out app through Ferry"\n'
initial_ingress = '''log "Initializing Cloudflare ingress before Ferry preflight"
ingress_payload="$(jq -nc --arg host "$APP_HOSTNAME" '{config:{ingress:[{hostname:$host,service:"http://dokku:80"},{service:"http_status:404"}]}}')"
cf_result "$(cf PUT "/accounts/$CF_ACCOUNT_ID/cfd_tunnel/$TUNNEL_ID/configurations" "$ingress_payload")" >/dev/null

log "Deploying the checked-out app through Ferry"
'''
if marker not in source:
    raise SystemExit("Deploy marker missing")
source = source.replace(marker, initial_ingress, 1)
path.write_text(source)
